library/ssl_tls13_client.c

```diff
@@ -1294,63 +1293,54 @@ static int ssl_write_key_shares_ext( mbedtls_ssl_context *ssl,
      MBEDTLS_SSL_DEBUG_MSG( 3, ( "Ciphersuites require a key length of max %d bits", max_size ) );
     */
 
-    /* The key_shares_curve_list provides us information about what we are expected to
-     * send, either based on the info provided by the app or by info offered by the server
-     * using the HRR.
-     */
-    nr = 0;
-
-    for ( grp_id = ssl->handshake->key_shares_curve_list; *grp_id != MBEDTLS_ECP_DP_NONE; grp_id++ ) {
-
-        info = mbedtls_ecp_curve_info_from_grp_id( *grp_id );
-
-        /* Check whether the key share matches the selected ciphersuite
-         * in terms of key length.
-         *
-         * Hence, AES-128 should go with a group bit size of 192 and 224 bits.
-         *
-         * TBD: Do we need this check?
-
-         switch ( max_size ) {
-         case 128: if( info->bit_size != 256 && info->bit_size!=192 && info->bit_size!= 224 ) continue;
-         break;
-         case 384: if( info->bit_size != 384 ) continue;
-         break;
-         }
-        */
+    grp_id = ssl->handshake->key_shares_curve_list;
```

Comment:
I only now realize that this is a different list from the supported curve list `ssl->conf->curve_list`.

Now that we only offer one share a time, I think we should remove `key_shares_curve_list`, and offer the first entry of `ssl->conf->curve_list` by default.

Reply:
If we want the user to be able to configure the use of a group different from the 0-th entry in the `curves` list, we could repurpose `mbedtls_ssl_conf_key_share_curves()` for that as `mbedtls_ssl_conf_tls13_default_group()`.

Reply:
So, my current recommendation would be the following:
* Remove `mbedtls_ssl_conf_key_shares_curves()`
* Remove `ssl->conf->key_shares_curve_list`
* Introduce a getter `ssl_tls13_get_key_share_group()` which is used in ClientHello to obtain the ID of the group for which a key share should be generated.
* As a first approximation, define `ssl_tls13_get_key_share_group()` as returning the first element of `ssl->conf->curves`.
* As a refinement, add a field `offered_key_share_group` and a configuration function `mbedtls_ssl_conf_tls13_default_group()` which allows users to set a different group ID than `curves[0]` for the initial key share.
* When we finalize support for HRRs requesting different groups, we can then re-set `offered_key_share_group` to whatever the server prefers, and the next `ClientHello` will use that.